merkle-tree/concurrent/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+use std::{cmp::Ordering, marker::PhantomData};
+
+use light_hasher::{errors::HasherError, Hasher};
+
+pub mod changelog;
+pub mod hash;
+
+use crate::{
+    changelog::ChangelogEntry,
+    hash::{compute_parent_node, full_proof, is_valid_proof},
+};
+
+pub type MerkleProof<const MAX_HEIGHT: usize> = [[u8; 32]; MAX_HEIGHT];
+
+#[repr(C)]
+pub struct ConcurrentMerkleTree<H, const MAX_HEIGHT: usize, const MAX_ROOTS: usize>
+where
+    H: Hasher,
+{
+    /// History of Merkle proofs.
+    pub changelog: [ChangelogEntry<MAX_HEIGHT>; MAX_ROOTS],
+    /// History of roots.
+    pub roots: [[u8; 32]; MAX_ROOTS],
+    /// Index of the newest changelog and root.
+    pub current_index: u64,
+    /// The newest Merkle proof.
+    pub rightmost_proof: MerkleProof<MAX_HEIGHT>,
+    pub rightmost_index: u64,
+    pub rightmost_leaf: [u8; 32],
+
+    _hasher: PhantomData<H>,
+}
+
+impl<H, const MAX_HEIGHT: usize, const MAX_ROOTS: usize> Default
+    for ConcurrentMerkleTree<H, MAX_HEIGHT, MAX_ROOTS>
+where
+    H: Hasher,
+{
+    fn default() -> Self {
+        Self {
+            changelog: [ChangelogEntry::default(); MAX_ROOTS],
+            roots: [[0u8; 32]; MAX_ROOTS],
+            current_index: 0,
+            rightmost_proof: [[0u8; 32]; MAX_HEIGHT],
+            rightmost_index: 0,
+            rightmost_leaf: [0u8; 32],
+            _hasher: PhantomData,
+        }
+    }
+}
+
+impl<H, const MAX_HEIGHT: usize, const MAX_ROOTS: usize>
+    ConcurrentMerkleTree<H, MAX_HEIGHT, MAX_ROOTS>
+where
+    H: Hasher,
+{
+    pub fn init(&mut self) {
+        // Initialize changelog.
+        let root = H::zero_bytes()[MAX_HEIGHT];
+        let mut changelog_path = [[0u8; 32]; MAX_HEIGHT];
+        for (i, leaf) in changelog_path.iter_mut().enumerate() {
+            *leaf = H::zero_bytes()[i];
+        }
+        let changelog_entry = ChangelogEntry::new(root, changelog_path, 0);
+        self.changelog[0] = changelog_entry;
+
+        // Initialize root.
+        self.roots[0] = root;
+
+        // Initialize rightmost proof and leaf.
+        for (i, leaf) in self.rightmost_proof.iter_mut().enumerate() {
+            *leaf = H::zero_bytes()[i];
+        }
+    }
+
+    pub fn root(&self) -> [u8; 32] {
+        self.roots[self.current_index as usize]
+    }
+
+    /// Finds the index of the current root in the tree's root buffer.
+    ///
+    /// # Use case and performance
+    ///
+    /// This method should be called in client implementations and tests. It
+    /// should not be actually used on-chain! It would regress the perfmance
+    /// of the on-chain tree by performing an unnecessary linear search.
+    pub fn root_index(&self) -> Result<usize, HasherError> {
```

Comment:
Why don't you just return current index?

Reply:
Lol, I'm absolutely dumb for doing this. :skull: :man_facepalming:  Thanks!